rnd/autogpt_server/autogpt_server/server/server.py

```diff
@@ -365,39 +246,64 @@ async def get_graphs(cls) -> list[str]:
 
     @classmethod
     async def get_templates(cls) -> list[autogpt_server.data.graph.GraphMeta]:
-        return await autogpt_server.data.graph.get_template_meta()
+        return await autogpt_server.data.graph.get_graphs_meta(is_template=True)
 
     @classmethod
-    async def get_graph(cls, graph_id: str) -> autogpt_server.data.graph.Graph:
-        graph = await autogpt_server.data.graph.get_graph(graph_id)
+    async def get_graph(
+        cls, graph_id: str, version: int | None = None
+    ) -> autogpt_server.data.graph.Graph:
+        graph = await autogpt_server.data.graph.get_graph(graph_id, version)
         if not graph:
             raise HTTPException(status_code=404, detail=f"Graph #{graph_id} not found.")
         return graph
 
     @classmethod
-    async def get_template(cls, graph_id: str) -> autogpt_server.data.graph.Graph:
-        graph = await autogpt_server.data.graph.get_graph(graph_id, is_template=True)
+    async def get_graph_history(
+        cls, graph_id: str
+    ) -> list[autogpt_server.data.graph.Graph]:
+        graphs = await autogpt_server.data.graph.get_graph_history(graph_id)
+        if not graphs:
+            raise HTTPException(status_code=404, detail=f"Graph #{graph_id} not found.")
+        return graphs
+
+    @classmethod
+    async def get_template(
+        cls, graph_id: str, version: int | None = None
+    ) -> autogpt_server.data.graph.Graph:
+        graph = await autogpt_server.data.graph.get_graph(graph_id, version)
         if not graph:
             raise HTTPException(status_code=404, detail=f"Graph #{graph_id} not found.")
+        if not graph.is_template:
+            raise HTTPException(
+                status_code=400, detail=f"Graph #{graph_id} is not a template."
+            )
         return graph
 
     @classmethod
     async def create_new_template(
         cls, create_graph: CreateGraph
     ) -> autogpt_server.data.graph.Graph:
-        if create_graph.graph:
-            graph = create_graph.graph
-        elif create_graph.tempalte_id:
-            graph = await autogpt_server.data.graph.get_graph(
-                create_graph.tempalte_id, is_template=True
-            )
-        else:
-            raise HTTPException(
-                status_code=400, detail="Either graph or template_id must be provided."
-            )
+        return await cls.create_graph(create_graph, is_template=True)
 
+    @classmethod
+    async def create_new_graph(
+        cls, create_graph: CreateGraph
+    ) -> autogpt_server.data.graph.Graph:
+        return await cls.create_graph(create_graph, is_template=False)
+
+    @classmethod
+    async def update_graph(
+        cls, graph: autogpt_server.data.graph.Graph
+    ) -> autogpt_server.data.graph.Graph:
         # TODO: replace uuid generation here to DB generated uuids.
-        graph.id = str(uuid.uuid4())
+        graph.graph_id = str(uuid.uuid4())
+        graph.version += 1
```

Comment:
This will break if an older version is retrieved, edited, and saved. We should probably find `max(version)` in the DB and increment from there. Or maybe we can let the DB do the incrementing natively?